experimental/dds/tree2/docs/schema2.md

```diff
@@ -0,0 +1,129 @@
+
+## Motivation
+
+The current schema system has several issues:
+
+1. It's cluttered with features that are not used in most cases. (ex: extraLocalFields, extraGlobalFields, global fields and values are not used on almost all schema)
+2. It's not clear to users which features when and why they all exist.
+3. global value fields and extra global fields don't interact in a clear well defined way (would such value fields be required on all nodes with extraGlobalFields or are they implicitly made optional when used via extra global fields?)
+4. Extra local fields can result in types TypeScript can't model, since our schema system only applies the schema for the extra fields to all fields not explicitly listed. This [can't be done in TypeScript](https://www.typescriptlang.org/play?noPropertyAccessFromIndexSignature=true&ts=4.5.5#code/PTAEBUCcE9QFwPagLYEMDWBTUr7QA7YDuAFppNgGYCWmANgCagBEqzo1AzqJ3JNQDsA5gBocAphV44KoAQFdkAI3IAuAFDqQoACIJM3AQjigiCSOnVwC2AMrVk+OtgC8oAN7qAkAG0ssQR4+QSEAXVVQf1BMAA84TAluVnYAfiD+YVAIhWVyAG51AF8CrTA9AzljU3N0DWtCUFtg4U4AeTgySHASVAEAHnBouISGbl4MoQA+UDd-BEoPUD9MAIFQAApBgB90kIBKcMiVofjEiFA0gUwAN3Iso9hiqxtG5qE2jvJu3oAmAZORmM3tM3ABRGIAYzo8gYmD64xCYnAkxK2gAQsYSPAyJxsKhZAAreTSAAGCOEJOeDVsM1eEw+nW+-WSKKpdh+tKa9PajJ6Aj+LNRZX0hiqSnk1EY6kE8UglFQELsDicmA5ni8qAi5KEBV8UUCnEOckUKkgBSepV0IsqJnFkoY0oEsvlisayucau8mt2wl1y1WjSNOVN5pKEIQAmkAH0GKg4KgAIwReyOZy09w4CLMGLsJ7hyMmGNx1A-ZPu1yLb3Z5hiJQRACsoCe2gAAnBOABaWKECFwLuQSDmUyDzL504meqYdT56Ox+MAZjLqYrGe9jebYDbne7mF7-cHkGHEaEoDHCQnNmnEdnxYALEuVenMywc7Ws9Bc7rNF4gA)
+5. It requires use of symbols to avoid colliding with local field names in many places in the API, which has been confusing users (many TypeScript developers don't even know what Symbols are).
+6. Supporting all the things in all the nodes makes APIs complex. This is extra annoying since this complicates the schema-aware API.
+7. Extra fields require proxies to implement the desired TypeScript APIs for nodes.
+8. Having both global and local fields adds complexity, particularly to schema and path storage where symbols can't be used.
+9. Supporting extraLocalFields and extraGlobalFields on nodes while providing a JavaScript object like API requires a Proxy, which adds overhead and complicates code implementation and maintenance.
+
+## Proposal
+
+Instead of one kind of node in view schema, have 4, each with a subset of our current functionality:
+
+1. Terminal Node: holds a value, but no children.
+   Don't support `undefined` as a value: an empty Struct can be used for that case if needed.
+2. Struct Node: finite list of fields (key+field type).
+3. Map Node: node with a single single field type which must permit empty.
+   Allows any field of that type under any string field key.
+   Provides the functionality currently done by making a node schema with extra local fields.
+4. Field Node: Has a single unnamed filed (using the empty field key).
```

Comment:
There are two concepts here:
1. A node that's parameterized with a field kind and implicitly unwraps it. I think "Transparent Node" may be a better name for this.
2. A specific instantiation of 1. Right now, it looks like sequence field is the only field that would benefit from this, so we'd likely call that instantiation "SequenceNode".

Reply:
There are use cases for doing this with optional fields: maybe not super high value usecases. For example, you could have a field which takes a color, and implement a color which can be transparent as an optional color node. This could be useful in a context where you want to take in that optional value or something else, like a button that that an optional color (which can be transparent) or an image.

I'm not sure Transparent Node is a great name since terminal nodes are transparent (and just show the value) by default, where field nodes do the same for a field.

Reply:
I'm not sure I fully follow the example. Describing it using TS types, are you saying that it lets us have

`OptionalColor = { [""]: Color | undefined }`
`ButtonAppearance = OptionalColor | Image`

instead of having

`OptionalColor = Color | undefined`
`ButtonAppearance = OptionalColor | Image` which is `Color | undefined | Image`

It's not clear to me in that example what value we get from the unwrapping.

Your point about the name "transparent" makes sense to me, and I don't have a better suggestion, so Field Node it is. I do think having `SequenceNode` be a thing and making sure most users interact with that concept without being forced to learn about `FieldNode` would be good, but I understand that's not this document's primary concern.

Reply:
Yes, thats the example. In that example, its different from having a single optional Image or Color since the optional color can have a merge identity even when its transparent, and can be edited to be not transparent. If you app has a color picker projected from the optional color for example that might work better. Also if image ws optional, and you wanted to be able to tell no image apart from transparent color, you would need some optional node pattern like this.

Reply:
I think that motivates having a wrapping layer. That part makes sense to me, it's the implicit unwrapping that doesn't make sense to me in this scenario.